### PR TITLE
LaunchScreen feature implementation

### PR DESCRIPTION
**Pull request by aszulak**
This commit is dependent on manifest-parser change:
https://review.tizen.org/gerrit/#/c/48846/

For more details:
https://crosswalk-project.org/documentation/manifest/launch_screen.html

### CODE REVIEW

File: runtime/browser/splash_screen.cc [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "runtime/browser/splash_screen.h"
+#include "runtime/browser/native_window.h"
+#include "common/logger.h"
```
Comment (WonyoungChoi) [inline runtime/browser/splash_screen.cc]: Wrong order of include. 
http://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Names_and_Order_of_Includes

File: runtime/browser/splash_screen.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+#define XWALK_RUNTIME_BROWSER_SPLASH_SCREEN_H_
+
+#include <manifest_handlers/splash_screen_handler.h>
+
+#include <map>
+#include <memory>
+#include <utility>
+#include <string>
+
+#include <Elementary.h>
```
Comment (WonyoungChoi) [inline runtime/browser/splash_screen.h:26]: Wrong order of include. 
http://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Names_and_Order_of_Includes

File: runtime/browser/splash_screen.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+class SplashScreen {
+ public:
+  SplashScreen(runtime::NativeWindow* window,
+               std::shared_ptr<const wgt::parse::SplashScreenInfo> ss_info,
+               const std::string& app_path);
+  void HideSplashScreen();
+
+ private:
+  wgt::parse::ScreenOrientation ChooseOrientation(
```
Comment (myroot) [inline runtime/browser/splash_screen.h]: make as static local function

File: runtime/browser/splash_screen.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      const std::map<wgt::parse::ScreenOrientation,
+                     wgt::parse::SplashScreenData>& splash_map,
+      runtime::NativeWindow::ScreenOrientation screen_orientation);
```
Comment (myroot) [inline runtime/browser/splash_screen.h]: 'runtime::NativeWindow::ScreenOrientation' has not been declared

You should include "native_widnow.h"

File: runtime/browser/splash_screen.cc [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  background_ = nullptr;
+  image_ = nullptr;
+  is_active_ = false;
+}
+
+std::pair<int, int> SplashScreen::GetDimensions() {
+  int w, h;
+#if defined(HAVE_X11)
+  uint16_t pid = getpid();
+  ecore_x_window_prop_property_set(elm_win_xwindow_get(window),
```
Comment (myroot) [inline runtime/browser/splash_screen.cc]: window variable was not declared in this function scope.

File: runtime/browser/splash_screen.h [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  std::pair<int, int> GetDimensions();
+  void SetBackground(const wgt::parse::SplashScreenData& splash_data,
+                     Evas_Object* parent, const std::pair<int, int>& pair,
```
Comment (myroot) [inline runtime/browser/splash_screen.h]: const std::pair<int, int>& pair 
variable name is not good. 
change to another name
(bound or size or ...)

File: runtime/browser/splash_screen.cc [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+  }
+  evas_object_show(background_);
+}
+
+void SplashScreen::SetImage(const wgt::parse::SplashScreenData& splash_data,
+                            Evas_Object* parent,
+                            const std::pair<int, int>& pair,
+                            const std::string& app_path) {
+  image_ = elm_image_add(background_);
+  elm_image_file_set(image_, (app_path + splash_data.image[0]).c_str(), NULL);
```
Comment (myroot) [inline runtime/browser/splash_screen.cc]: splash_data.image[0] 

What this mean?
unreadable code.

File: runtime/browser/splash_screen.cc [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+  } else if (!splash_data.background_color.empty()) {
+    elm_bg_color_set(background_, splash_data.color.red,
+                     splash_data.color.green, splash_data.color.blue);
+  }
+  evas_object_show(background_);
+}
+
+void SplashScreen::SetImage(const wgt::parse::SplashScreenData& splash_data,
+                            Evas_Object* parent,
+                            const std::pair<int, int>& pair,
```
Comment (myroot) [inline runtime/browser/splash_screen.cc]: I think int witdh,int height better than pair<int,int>

const std::pair<int, int>& pair ->
int width, int height

File: runtime/browser/splash_screen.cc [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+      runtime::NativeWindow::ScreenOrientation::PORTRAIT_PRIMARY) {
+    orientation_pair = splash_map.find(wgt::parse::ScreenOrientation::PORTRAIT);
+    if (orientation_pair == splash_map.end())
+      orientation_pair = splash_map.find(wgt::parse::ScreenOrientation::AUTO);
+  } else {
+    orientation_pair =
+        splash_map.find(wgt::parse::ScreenOrientation::LANDSCAPE);
+    if (orientation_pair == splash_map.end())
+      orientation_pair = splash_map.find(wgt::parse::ScreenOrientation::AUTO);
+  }
```
Comment (myroot) [inline runtime/browser/splash_screen.cc]:
```
if (orientation_pair == splash_map.end())
  orientation_pair = splash_map.find(wgt::parse::ScreenOrientation::AUTO);
```

this code is duplicated.

File: runtime/browser/web_application.cc [hunk 10]
```diff
@@ -683,15 +652,15 @@ void WebApplication::OnRendered(WebView* /*view*/) {
   STEP_PROFILE_END("URL Set -> Rendered");
   STEP_PROFILE_END("Start -> Launch Completed");
   LOGGER(DEBUG) << "Rendered";
+  splash_screen_->HideSplashScreen();
```
Comment (myroot) [inline runtime/browser/web_application.cc]: It is not good position to hide splash screen.

**Comment by myroot** [comment]
Please do feasibility test.
It seem to does not tested.

File: runtime/browser/splash_screen.h (hunk 2 above)
Comment (aszulak) [inline runtime/browser/splash_screen.h:26]: Fixed by changing include order.

File: runtime/browser/splash_screen.h (hunk 3 above)
Comment (aszulak) [inline runtime/browser/splash_screen.h]: I've moved this method to anonymous namespace in .cc file.

File: runtime/browser/splash_screen.h (hunk 4 above)
Comment (aszulak) [inline runtime/browser/splash_screen.h]: Fixed by including "native_window.h" header file.

File: runtime/browser/splash_screen.cc (hunk 5 above)
Comment (aszulak) [inline runtime/browser/splash_screen.cc]: Fixed by using variable passed in constructor of SplashScreen.

File: runtime/browser/splash_screen.h (hunk 6 above)
Comment (aszulak) [inline runtime/browser/splash_screen.h]: I renamed this variable to 'bound', which is more appropriate than 'pair'.

File: runtime/browser/splash_screen.cc (hunk 7 above)
Comment (aszulak) [inline runtime/browser/splash_screen.cc]: Slightly refactored this code to be more readable.

File: runtime/browser/splash_screen.cc (hunk 8 above)
Comment (aszulak) [inline runtime/browser/splash_screen.cc]: I do not agree - width and height represents a logical pair - thus this type is used.

**Comment by WonyoungChoi** [comment]
Please do not make additional commits like "Resolved code review ... ". 
You can use "git commit --amend" to change your last commit. 
Please refer below commands. 
$ cd (your local repo)
$ git reset 37623778983f260c7517be9c1a4dad7aca4ed543
$ git add (changed files)
$ git commit --amend
$ git push --force origin launch_screen

And below blog post can be helpful to write the git commit message. 
http://chris.beams.io/posts/git-commit/

**Comment by aszulak** [comment]
@WonyoungChoi Thank you for your help. I wasn't sure if I should use the same work flow like on gerrit, or push changes as a new commit.

File: runtime/browser/splash_screen.cc (hunk 9 above)
Comment (aszulak) [inline runtime/browser/splash_screen.cc]: Removed a duplicated code.

File: runtime/browser/web_application.cc (hunk 10 above)
Comment (aszulak) [inline runtime/browser/web_application.cc]: Could you recommend a better place where splash screen should be hidden?

**Comment by myroot** [comment]
You should rebase. 
Your commit is including many others changes unrelated files

File: common/resource_manager.cc [hunk 11]
```diff
@@ -243,7 +243,7 @@ std::unique_ptr<ResourceManager::Resource> ResourceManager::GetMatchedResource(
   if (!app_control_info.src().empty()) {
     return std::unique_ptr<Resource>(new Resource(
       InsertPrefixPath(app_control_info.src()),
-                       app_control_info.onreset() == "disable" ? false : true));
+                       app_control_info.reload() == "disable" ? false : true));
```
Comment (myroot) [inline common/resource_manager.cc:246]: It already changed by 

commit 281d5a86720db88bda1af1beeb6ff076d7bef759
Author: psikorski psikorski@users.noreply.github.com
Date:   Mon Oct 19 11:09:04 2015 +0200

```
Revert "Revert "config.xml parameter change: on-reset -> reload""
```

File: common/resource_manager.cc [hunk 12]
```diff
@@ -583,7 +583,7 @@ std::string ResourceManager::DecryptResource(const std::string& path) {
   rewind(src);
 
   // Read buffer from the source file
-  std::unique_ptr<char> src_buf(new char[src_len]);
+  std::unique_ptr<char[]> src_buf(new char[src_len]);
```
Comment (myroot) [inline common/resource_manager.cc:586]: It already changed by

commit 218a64dae6d8e57bbda7cea732fd972c26a1764d
Author: sung-su.kim sung-su.kim@samsung.com
Date:   Thu Oct 22 09:38:27 2015 +0900

```
Fix incorrectly unique_ptr declared

from unique_ptr<char> to unique_ptr<char[]>
```

**Comment by myroot** [comment]
This changes did not meet the requirement.

The splash screen should hide at the user selected time.
ready_when: 
- first-paint: The first visually non-empty paint has occurred. 
- user-interactive: The DOM and CSSOM have been constructed.
- complete: All the resources have been loaded.
- custom: The window.screen.show() method was called.

The background and image properly resize.

 image_border = "30px [30px 30px 30px] [repeat | stretch | round] [repeat | stretch | round]" />

image_border: A border image, split into a 9-piece image with the intermediate border pieces scaled and/or stretched to fill the whole viewport. See the Image Border spec(https://docs.google.com/document/d/17PuNuHRTQuREUpaCvj-eEx7uYi2avd-VW-oaMXMpvwo/edit?pli=1#heading=h.rq1ayw778vp6) for full details.

File: runtime/browser/web_application.cc (hunk 10 above)
Comment (myroot) [inline runtime/browser/web_application.cc]: first-paint: The first visually non-empty paint has occurred.
  -> WebApplication::OnRendered
complete: All the resources have been loaded.
  -> WebApplication::OnLoadFinished
 user-interactive: The DOM and CSSOM have been constructed.
 -> WebApplication::OnLoadFinished

custom: The window.screen.show() method was called.
 -> Should be implemented.

File: runtime/browser/splash_screen.cc [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+  auto orientation_pair = splash_map.end();
+
+  if (screen_orientation ==
+      runtime::NativeWindow::ScreenOrientation::PORTRAIT_PRIMARY) {
+     orientation_pair =
+         splash_map.find(wgt::parse::ScreenOrientation::PORTRAIT);
+  } else {
+    orientation_pair =
+        splash_map.find(wgt::parse::ScreenOrientation::LANDSCAPE);
+  }
```
Comment (myroot) [inline runtime/browser/splash_screen.cc:53]:
```
auto primary_orientation = screen_orientation == runtime::NativeWindow::ScreenOrientation::PORTRAIT_PRIMARY ? 
     wgt::parse::ScreenOrientation::PORTRAIT : wgt::parse::ScreenOrientation::LANDSCAPE;
orienatation_pair = splash_map.find(primary_orientation);
```

**Comment by WonyoungChoi** [comment]
Please rebase your PR commit. 
$ cd (your local repo)
$ git remote add xwalk-tizen https://github.com/crosswalk-project/crosswalk-tizen
$ git fetch xwalk-tizen
$ git rebase xwalk-tizen/master
$ git push -f origin launch_screen

**Comment by aszulak** [comment]
I've fixed one of the issues reported by @myroot, regarding the place where the splash screen is being hidden. I've tested it and it works - the missing feature is image border and scaling (basic scaling works as well). I've asked one of our software engineers, and the image border and scaling will require 3~5 days, thus I've marked this as TODO in this pull request.

File: runtime/browser/splash_screen.h [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+ private:
+  std::pair<int, int> GetDimensions();
+  void SetBackground(wgt::parse::SplashScreenData& splash_data,
+                     Evas_Object* parent, const SplashScreenBound& bound,
+                     const std::string& app_path);
+  void SetImage(const wgt::parse::SplashScreenData& splash_data,
+                Evas_Object* parent, const SplashScreenBound& bound,
+                const std::string& app_path);
+<<<<<<< HEAD
+=======
```
Comment (myroot) [inline runtime/browser/splash_screen.h]: It is not merged

File: runtime/browser/splash_screen.cc [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+#include "runtime/browser/native_window.h"
+
+namespace {
+
+wgt::parse::ScreenOrientation ChooseOrientation(
+    const std::map<wgt::parse::ScreenOrientation,
+    wgt::parse::SplashScreenData>& splash_map,
+    runtime::NativeWindow::ScreenOrientation screen_orientation) {
+  auto orientation_pair = splash_map.end();
+
```
Comment (myroot) [inline runtime/browser/splash_screen.cc:45]: using ScreenOrientation = runtime::NativeWindow::ScreenOrientation;

File: runtime/browser/splash_screen.h [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+class SplashScreen {
+  typedef std::pair<int, int> SplashScreenBound;
+ public:
+  SplashScreen(NativeWindow* window,
+               std::shared_ptr<const wgt::parse::SplashScreenInfo> ss_info,
+               const std::string& app_path);
+  void HideSplashScreen();
+=======
+enum class HideReason { RENDERED, LOADFINISHED };
```
Comment (myroot) [inline runtime/browser/splash_screen.h]: Move to SplashScreen inner enum class

**Comment by myroot** [comment]
Please change the name of extension for splash screen. 
Utils -> splash

**Comment by myroot** [comment]
Screen is not good. 
Please change to SplashScreen or Splash. 
You must include "Splash" keyword in extension name

**Comment by aszulak** [comment]
@myroot I've renamed 'Screen' to 'SplashScreen' as you requested.

File: extensions/internal/splash_screen/splash_screen.json [hunk 17]
```diff
@@ -0,0 +1,7 @@
+[
+  {
+    "name":"Screen",
```
Comment (myroot) [inline extensions/internal/splash_screen/splash_screen.json]: extension name was not changed. 
It should be same with 
setted name by below code
g_core->SetExtensionName(g_xw_extension, "SplashScreen");

File: extensions/internal/splash_screen/splash_screen_extension.cc [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+#include "common/string_utils.h"
+
+XW_Extension g_xw_extension = 0;
+const XW_CoreInterface* g_core = NULL;
+const XW_MessagingInterface* g_messaging = NULL;
+const XW_Internal_SyncMessagingInterface* g_sync_messaging = NULL;
+const XW_Internal_EntryPointsInterface* g_entry_points = NULL;
+const XW_Internal_RuntimeInterface* g_runtime = NULL;
+
+extern const char kSource_utils_api[];
```
Comment (myroot) [inline extensions/internal/splash_screen/splash_screen_extension.cc]: It should be 
extern const char kSource_splash_screen_api[];

**Comment by myroot** [comment]
PLEASE, Test before upload.

**Comment by aszulak** [comment]
@myroot It should be OK now.

**Comment by myroot** [comment]
LGTM

**Comment by WonyoungChoi** [comment]
I'm waiting for https://review.tizen.org/gerrit/#/c/48846/.

**Pull request merged**